lib/engine/src/engine.rs

```diff
@@ -101,4 +97,41 @@ impl dyn Engine {
             None
         }
     }
+
+    /// Downcast a dynamic Executable object to a concrete implementation of the trait.
+    pub fn downcast_arc<T: Engine + 'static>(self: Arc<Self>) -> Result<Arc<T>, Arc<Self>> {
```

Comment:
So the problem with `Arc::downcast` is that it is implemented for `Arc<dyn Any + Send + Sync + 'static>`. This notably doesn’t work for `Arc<dyn Engine*>` and there isn’t an easy way to go between the two, so you cannot actually call `Arc::downcast` on `Arc<dyn Engine>` AFAICT.

Reply:
Hmm I think having something like this should work?
```rust
trait Engine: Any + Send + Sync + 'static {
    // ...
    fn to_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync + 'static> {
        self
    }
}
```
And then use `Arc::downcast(engine.to_any())`?

Basically I'm feeling that if there's no way something like this would work, it might mean that this bit of unsafe code, however sound it looks, is actually wrong :/

Also I actually just noticed that `Engine::type_id` is a function implemented by a non-`unsafe` non-sealed trait, which means that the code is technically unsound because an engine could spoof a fake TypeId. That said tracing has similar unsoundness in its traits and it's not a big deal in practice, so it's probably not too bad.

Reply:
> Also I actually just noticed that Engine::type_id is a function implemented by a non-unsafe non-sealed trait, which means that the code is technically unsound because an engine could spoof a fake TypeId.

Implementers wouldn’t be able to access `private::Internal` necessary to specify the argument type for the function.

> And then use Arc::downcast(engine.to_any())?

Implementing `to_any` would require an unstable feature (trait upcasting) or `unsafe` code along the lines of what is seen here, and would end up returning `dyn Any` in its `Err` variant, which is not great in terms of usability of the API, but it would work.

Reply:
Hmm so for to_any(), here is what I was thinking of: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7cedaa843727797e370656ec18459e0f (found by re-looking at the craziest trait object code I ever wrote, [erlust](https://github.com/Ekleog/erlust))

I had forgotten that a default impl of the function wouldn't work, but that's probably not a problem for us. The advantage being, that there's zero unsafe code here. And the drawback that, as you mentioned, it returns `dyn Any` instead of `dyn Engine` upon failure, but that's probably not too bad as it's all `Arc` anyway and we can just keep a clone of the `Arc<dyn Engine>` from before the `to_any` cast.

WDYT?